firecracker-control/local.go

```diff
@@ -0,0 +1,319 @@
+// Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package service
+
+import (
+	"crypto/rand"
+	"fmt"
+	"math"
+	"math/big"
+	"os"
+	"path/filepath"
+	"strconv"
+	"sync"
+
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/containerd/plugin"
+	"github.com/firecracker-microvm/firecracker-go-sdk"
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+
+	"github.com/firecracker-microvm/firecracker-containerd/proto"
+)
+
+var (
+	_ proto.FirecrackerServer = (*local)(nil)
+)
+
+func init() {
+	plugin.Register(&plugin.Registration{
+		Type: plugin.ServicePlugin,
+		ID:   localPluginID,
+		InitFn: func(ic *plugin.InitContext) (interface{}, error) {
+			log.G(ic.Context).Debugf("initializing %s plugin (root: %q)", localPluginID, ic.Root)
+			return newLocal(ic)
+		},
+	})
+}
+
+// TODO: Remove this interface after merging PR #91
+// See https://github.com/firecracker-microvm/firecracker-go-sdk/pull/91
+type machineIface interface {
+	Start(context.Context) error
+	StopVMM() error
+	Wait(context.Context) error
+	SetMetadata(context.Context, interface{}) error
+}
+
+type instance struct {
+	cfg     *firecracker.Config
+	machine machineIface
+}
+
+type local struct {
+	vm          map[string]*instance
+	vmLock      sync.RWMutex
```

Comment:
In general I prefer map+lock because of type safety and use sync.Map for performance heavy situations. We'll remove lock/unlock headache and will add interface casting headache instead.